#29 Block sparse GEMM implementation

mzuzek opened the pull request:
Fixes #29

Mirror PR in the upstream repository: https://github.com/kokkos/kokkos-kernels/pull/1099

cwpearson committed:
Experimental tensor-core SpMV for BsrMatrix.

Extensible to any CUDA wmma fragment type and any BsrMatrix::blockDim.
Each team is made up of four warps, which are in a 2x2 grid in the M and
N direction of the sparse matrix.
The input operands are all operated on at a granularity defined by the
BsrMatrix blockDim.
Each dense block in the product MV is covered by a 2D grid of teams as
needed, where the footprint of each team is determined by the chosen
fragment type.

src/sparse/KokkosSparse_BlockCrsMatrix.hpp
====
Added an is_block_crs_matrix static test, similar to the is_view from Kokkos.
This is not actually used in this PR, but is convenient for future
KokkosSparse::spmv extensions.

src/sparse/KokkosSparse_CrsMatrix.hpp
====
Added an is_crs_matrix static test, similar to the is_view from Kokkos.
This is not actually used in this PR, but is convenient for future
KokkosSparse::spmv extensions.

src/sparse/KokkosSparse_spmv.hpp
====
Integration with the existing SpMV interface.
The current KokkosSparse::spmv(controls, mode, ...) interface is
extended with compile-time checks for whether the AMatrix type is a
BsrMatrix or not. If so, the new implementation is called.
If it's a CrsMatrix, the existing implementation is called.
If neither, a stub implementation that provides a useful error is
introduced.
Prior to this PR, passing a non-CrsMatrix to Kokkos::spmv would be a compile-time error of varying transparency.
The user must provide a control with "algorithm" =
"experimental_bsr_tc", or the spmv will be a run-time error.
Furthermore, on AMPERE, the internal fragment type may be controlled
with "precision" = "mixed" or "double".

src/sparse/impl/KokkosSparse_spmv_bsrmatrix_impl.hpp
====
The actual tensor core SpMV implementation.
In the future, this file will hold additional impls for the BsrMatrix
type.
The functor has optional compile-time parameters that can be used to
optimize out divmod instructions.
If they are provided as 0, these parameters are determined at runtime
instead.
The entire implementation is guarded by the Kokkos VOLTA architecture with CUDA enabled.
FP16,32, and 64 are supported on any inputs/outputs. The multiplicand fragments are staged through shared memory where they are converted to the internal fragment type.

unit_test/sparse/Test_Sparse_spmv.hpp
====
Added unit test for the new SpMV.
It is invoked through the same opt-in interface that is exposed to the user.
It is compared against a Kokkos::spmv() called on the same operands

src/CMakeLists.txt
====
Add in ETI files for the BsrMatrix spmv

src/impl/generated_specializations...
====
ETI files for the BsrMatrix spmv

brian-kelley committed:
Implement batched serial SVD

- based on Golub/van Loan
- supports just float/double now
- full SVD or just singular values

e10harvey committed:
src/batched: Start adding DblBufGemm

e10harvey committed:
src/batched: Fix build, start adding policy

e10harvey committed:
src/batched: Fix BatchedGemm tests

e10harvey committed:
src/batched: Start adding functor

e10harvey committed:
src/batched: Fix team_policy runtime errors

lucbv committed:
Blas GEMM: fix early exit logic, see issue #1088

The GEMM algorithm should perform a quick return
if C is empty, otherwise if A and/or B are empty
a simple scaling of C by beta can be performed
before exiting.

e10harvey committed:
src/batched: initial DblBufGemm implementation

e10harvey committed:
src/batched: cleanup and TODOs

e10harvey committed:
src/batched: formatting

e10harvey committed:
src/batched: formatting

lucbv committed:
Blas GEMM: adding new sizes to test to cover bug from issue #1088

Adding a test with sizes 20, 14, 0 to check correct implementation
of the quick return when A.extent(1) == 0.
Alos fixing test to have it exercise the quick return code path.

e10harvey committed:
src/batched: Support transpose in DblBufGemm

  - For Kokkos::ALL(), subview_wrapper did not transpose
  the view properly. These changes add a transpose_2d_view
  utility function used for modifying the underlying strides
  and extents of the view.
  - DblBufGemm nows accounts for transpose args when initializing
  k.
  - Cleanup unittests.

e10harvey committed:
src/batched:

  - Cleanup from self review
  - Move Internal APIs to Impl
  - Add macro for conditional unrolling

e10harvey committed:
src/batched: Add TODO

e10harvey committed:
src/batched: Checkpoint

  - Fix invalid cuda addr
  - Fix transpose bugs with BatchLayout::Left (LayoutRight)

e10harvey committed:
src/batched: Checkpoint

  - Fix subview wrappers for vectors & scalars

e10harvey committed:
perf_test/blas/blas3: Add BatchedGemm perf test

mzuzek commented:
@uhetmaniuk @fnrizzi 

I got CPU version of the default algorithm working:
* Symbolic version is reused from non-block `spgemm` without modification.
* Numeric phase was implemented `KokkosBSPGEMM` in _sparse/impl/KokkosSparse_bspgemm_impl_kkmem.hpp_ with `MultiCoreTag` (`SPGEMM_KK`) and `MultiCoreTag4`  (`SPGEMM_KK_LP`) functors.
* Accumulator is work-in-progress: `MultiCoreTag` operator uses `HashingIndex` which is essentially `HashmapAccumulator` reduced to indexing used by spgemm (`HashmapAccumulator` works on scalars) while `MultiCoreTag4` operator inlines the indexing like it did in non-block version. I'll try to refactor/unify it once I dig into GPU operators and understand all indexing/accumulation scenarios. This plan includes detailed review of related memory estimations and allocations.
* For block accumulation/combination I used `KokkosBatched::SerialGemmInternal<KokkosBatched::Algo::Gemm::Unblocked>`  from _sparse/impl/KokkosBatched_Gemm_Serial_Internal.hpp_ (`C = A * B` and `C += A * B` with `beta` of 0 and 1 respectively).
* Unit tests are based on new overload of `kk_generate_sparse_matrix()`  in _common/KokkosKernels_IOUtils.hpp_ that takes block size and generates random BSR matrix (random CSR and BSR content is generated by shared `generate_bsr_content()` routine, I also fixed couple of corner cases there, so it can generate small matrices as well).
* There are some extra utilities for unit testing in _unit_test/sparse/Test_Sparse_bspgemm.hpp_, including `<<` operator that renders BSR matrix on `stdout`.

Edit Sep 9, 2021: added note on batched dgemm for block accumulation

mzuzek commented:
@uhetmaniuk @fnrizzi 

I connected MKL to benchmark block spgemm performance on CPU.

> Note: In non-block version, `mkl_sparse_spmm()`  is called twice: in symbolic and numeric phases. I drafted a new interface that avoids this duplicated allocations and calculation to get more accurate insight into performance baseline. This optimization would need some discussion to be included in the actual implementation (i.e. outside the unit test), since it doesn't immediately fit the interface, which forces separate symbolic and numeric phases.

For multiplication of random matrix by itself on [i9-9900K](https://www.intel.com/content/www/us/en/products/sku/186605/intel-core-i99900k-processor-16m-cache-up-to-5-00-ghz/specifications.html) (8 cores/16 threads)/32GB RAM/Ubuntu/gcc-8.4 (WSL), I got some rough timings (no repetition, accuracy around 0,5~1s):

```
Note: Google Test filter = *block_spgemm*
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from threads
[ RUN      ] threads.sparse_block_spgemm_double_int_int_TestExecSpace

Testing BSPGEMM: nrows=300000   block:2x2       nnz=8000000(0.009%)     bandwidth=100000(33.3%)        variance=10 (0.0%)      schedule=dynamic
 nnz_in=7800201
nnz_out=202326755
SPGEMM_DEBUG (reference result): 17.0
SPGEMM_MKL              spgemm_time:    5.5     output_check_time:      0.6
SPGEMM_KK               spgemm_time:    1.7     output_check_time:      0.5
SPGEMM_KK_LP            spgemm_time:    1.9     output_check_time:      0.5

Testing BSPGEMM: nrows=300000   block:3x3       nnz=6000000(0.007%)     bandwidth=100000(33.3%)        variance=10 (0.0%)      schedule=dynamic
 nnz_in=6000201
nnz_out=119850101
SPGEMM_DEBUG (reference result): 23.0
SPGEMM_MKL              spgemm_time:    10.7    output_check_time:      6.7
SPGEMM_KK               spgemm_time:    4.8     output_check_time:      0.7
SPGEMM_KK_LP            spgemm_time:    2.2     output_check_time:      0.7
[       OK ] threads.sparse_block_spgemm_double_int_int_TestExecSpace (80583 ms)
[----------] 1 test from threads (80587 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (80589 ms total)
[  PASSED  ] 1 test.
```
> Note: I tried to measure largest matrices I could fit into memory. For reference: _audikw_1_ - our largest [_spmv_ sample](https://github.com/fnrizzi/kokkos-kernels/issues/3#issuecomment-842759224) has around 40e+6 non-zero values.

> Note: _nnz_ means non-zero **blocks**.

> Note: Tested on _Threads_ Kokkos execution space (because of silly issue with _OpenMP_ running on single core...)

Would that look reasonable - based on your spgemm perf experience/expectations ?
Or should we run more tests ?

uhetmaniuk commented on src/common/KokkosKernels_IOUtils.hpp:
```diff
@@ -413,61 +412,113 @@ crsMat_t kk_generate_triangular_sparse_matrix(
 }
 
+template <typename bsrMat_t>
+bsrMat_t kk_generate_sparse_matrix(
```
Naive question: could you fill first a CrsMatrix and then use the BsrMatrix constructor that takes a CrsMatrix and a block dimension as input parameters?

mzuzek replied:
Yes, I reverted to this simpler approach

uhetmaniuk commented:
Looks good. I do not have any other comments.

e10harvey committed:
src/batched: Fix CI & code cleanup

e10harvey committed:
src/batched: Improve LL perf

  - Improve shmem access pattern
  - Update SQUARE heuristics

brian-kelley committed:
Fix #1089 (failing batched UTV tests)

by adding missing team barriers

kliegeois committed:
Move the batched dense files to specific directories

kliegeois committed:
Add batched CG

ndellingwood committed:
Merge pull request #1096 from brian-kelley/Fix1089

Fix #1089 (failing batched UTV tests)

mzuzek commented:
@uhetmaniuk @fnrizzi 
Thank you for the review. I opened draft [PR#1099](https://github.com/kokkos/kokkos-kernels/pull/1099) in the KK repository.

cwpearson committed:
Allow reading integer mtx files into floating-point matrices

lucbv committed:
Merge pull request #1091 from lucbv/Fix_gemm_shortcut

Blas GEMM: fix early exit logic, see issue #1088

brian-kelley committed:
Added test for C's dimensions in spadd

(simplified interface)

brian-kelley committed:
Fix C's dimensions in spadd simplified interface

Also skips the "maximum_entry" kernel in CrsMatrix, which takes O(nnz)

kliegeois committed:
Merge remote-tracking branch 'origin/develop' into batched-dense

lucbv committed:
Remove unused type definition

As we switch from graph constructor to view constructor we do not need the graph type anymore

lucbv committed:
Merge pull request #1100 from cwpearson/fix/read-integer-mtx-into-fp

Allow reading integer mtx files into floating-point matrices

lucbv committed:
Merge pull request #1102 from brian-kelley/FixSpaddColumns

Fix C's numCols in spadd simplified interface

lucbv committed:
Merge pull request #1098 from kliegeois/batched-dense

Move the batched dense files to specific directories

e10harvey committed:
src/batched: Update DblBuf comments

perf_test/blas/blas3:
  - Refactor gemm fn names for consistency and clarity
  - Add gemm csv output options
  - Register finalize with exit

e10harvey committed:
src/batched: Move DblBuf to dense/impl sub-dir

e10harvey committed:
src/batched: Fix build warning

brian-kelley committed:
Merge pull request #1107 from brian-kelley/BatchedSVD

Batched serial SVD

e10harvey committed:
src/batched/dense: Add KK_SERIAL_RANK0

brian-kelley committed:
Fix #1111: spmv tpl instantiations

Use KOKKOS_COMPILE_LIBRARY (true while building library, and false when external app is using prebuilt library)
as the "eti avail" template parameter for SpMV TPL definitions (cuSPARSE and MKL). Before was hardcoded to true,
but this caused undefined reference errors when the TPL was available but ETI was not.

lucbv committed:
Merge pull request #1112 from brian-kelley/Fix1111

Fix #1111: spmv tpl instantiations

lucbv committed:
CrsMatrix: deprecate constructor without ncols input

We have only one constructor that does not accept
ncols as an input (except for default constructor)
Additionally that constructor does extra work to
compute that value without really telling the user.
The changes deprecate that behavior and move some
code around to group more logically attributes and
methods in the class.

lucbv committed:
CrsMatrix: small fix in test to use new contructor not deprecated one.

e10harvey committed:
docs: Add initial doxygen generation

lucbv committed:
Merge pull request #1115 from lucbv/CrsMatrix_graph_constructor

CrsMatrix: deprecate constructor without ncols input

e10harvey committed:
docs: Add initial sphinx generation

e10harvey committed:
docs: Include Kokkos, Sparse, Blas, and Batched namespace members

e10harvey committed:
cmake: Add KokkosKernels_ENABLE_DOCS

  - Point docs to Kokkos include dir

brian-kelley committed:
Fix SpGEMM for Nvidia Turing/Ampere

by enabling the code paths required for "independent thread scheduling"
that was introduced with Volta.

brian-kelley committed:
Merge pull request #1118 from brian-kelley/FixSpgemmAmpere

Fix SpGEMM for Nvidia Turing/Ampere

e10harvey committed:
perf_test/blas/blas3: Add benchmarking script for reproducability

e10harvey committed:
Merge pull request #1095 from e10harvey/issue1045

src/batched: Add BatchedDblBufGemm

e10harvey committed:
Merge pull request #1116 from e10harvey/automate_docs

Automate documentation generation

e10harvey committed:
Address feedback from #1095

e10harvey committed:
Fixes #1121

e10harvey committed:
example/gmres: Fix XL build

e10harvey committed:
perf_test/blas/blas3:
  - Fix --csv option in KokkosBlas3_perf_test
  - Update benchmark script:
    - reduce build time
    - fix typos
    - add comments
    - wait for file system writes to land

e10harvey committed:
perf_test/blas/blas3:
  - Small refactor in benchmark script

e10harvey committed:
perf_test/blas/blas3:
  - Add A64FX and SKX to benchmark script
  - Add --use_simd option to perf_test

e10harvey committed:
src/batched:
  - Add simd support for KK_SQUARE.

 perf_test/blas/blas3:
   - pass simd views for batched heuristic.

e10harvey committed:
Merge pull request #1122 from e10harvey/issue1121

Fix BatchedDlbBuf build with cuda-9 and XL toolchains

ndellingwood committed:
gmres: disable examples for builds with ibm/xl

Related to https://github.com/kokkos/kokkos/issues/4376

ndellingwood committed:
Merge pull request #1123 from ndellingwood/disable-gmres-xlc

gmres: disable examples for builds with ibm/xl

e10harvey committed:
src/batched:
  - Add debug prints for KK_SQUARE.

 perf_test/blas/blas3:
   - Use srun on SKX and A64FX.

ajpowelsnl committed:
feature/gemv rps test -- RAJAPerf Suite Version of the BLAS2 GEMV Test (#1085)

Initial RAJAPerf Suite (RPS)-based performance testing.

Dot (device-level and team), GEMV, SpMV.
Adding RAJAPerf Suite as a git submodule.
Enable RPS with -DKokkosKernels_ENABLE_TESTS_AND_PERFSUITE=ON.

See PRs #1085 and #1037 for more info.

Co-authored-by: David Poliakoff <david.poliakoff@gmail.com>
Co-authored-by: Amy Powell <ajpowel@.gov>
Co-authored-by: brian-kelley <bmkelle@sandia.gov>

ndellingwood committed:
Workaround error with intel

Providing explicit template args to some calls to TeamGemv::invoke within the supernodal sptrsv helps workaround compiler issue deducing the types
Occurs with Intel builds of Trilinos

ndellingwood committed:
Rename CUDA_SAFE_CALL KOKKOS_IMPL_CUDA_SAFE_CALL

Change to match kokkos/kokkos#4249

ndellingwood committed:
Merge pull request #1128 from ndellingwood/intel-workaround

Workaround error with intel

ndellingwood committed:
Merge pull request #1130 from ndellingwood/rename-cudasafecall

Rename CUDA_SAFE_CALL to KOKKOS_IMPL_CUDA_SAFE_CALL

lucbv committed:
Stream interface: adding stream support in GEMM

This will allow users to queue up kernels in individual
stream when the GPU cannot be fully utilized by a single
kernel call. This might happen if the matrices are not
so large to occupy the full GPU but also not small enough
to be batched.

ajpowelsnl committed:
blas3: rps version of gemm test

ajpowelsnl committed:
Addressing BMK PR#1132 comments: all comments accepted

kliegeois committed:
Add batched axpy and batched spmv (#1092)

* Add batched dense axpy and batched spmv

lucbv committed:
Stream interface: adding gemv interface to support gemv_based_gemm

This will probably now require some more testing gemv and maybe
 also additional testing in gemm to cover both gemv_based_gemm
 and dot_based_gemm.

e10harvey committed:
Merge pull request #1120 from e10harvey/issue933

perf_test/blas/blas3: Add benchmarking script for reproducability

e10harvey committed:
Revert "gmres: disable examples for builds with ibm/xl"

This reverts commit adaa5510f9db1c9fbac8fef8e13afa6f23f0d383.

e10harvey committed:
Merge pull request #1132 from ajpowelsnl/feature/gemm_PR

RAJAPerf Suite Version of the Blas 3 gemm test -- gemm changes only

e10harvey committed:
Apply clang-format to code base.

.github/workflows: Add clang-format-check action

e10harvey committed:
Merge pull request #1134 from e10harvey/kokkos_issue4376

Revert "gmres: disable examples for builds with ibm/xl"

e10harvey committed:
Merge pull request #1135 from e10harvey/issue1002

Apply clang-format and add action to enforce coding style

e10harvey committed:
.github/workflows: Remove test comment

lucbv committed:
Stream interface: clean-up some cublas TPL calls

Adding a new macro KOKKOS_CUBLAS_SAFE_CALL_IMPL
to check error codes returned by cublas API.
If the error code is not CUBLAS_STATUS_SUCCESS,
the macro prints the error type and message as
well as the filename and line where the error
was detected and throws a runtime exception.

cwpearson committed:
use row_map_type instead of index_type for row_map

e10harvey committed:
Merge pull request #1137 from e10harvey/remove-test-comment

.github/workflows: Remove test comment

e10harvey committed:
perf_test/blas/blas3: Update SHAs for benchmarking

e10harvey committed:
.github/workflows: Fix regex

lucbv committed:
Deprecation warning: SpaceAccessibility move out of impl, see #1140

Kokkos::impl::SpaceAccessibility moved out of the impl namespace
in kokkos/kokkos#1115 and was deprecated in favor of
Kokkos::SpaceAccessibility. This PR removes all calls to the
deprecated version.

lucbv committed:
Merge pull request #1138 from cwpearson/fix/issue-1133

Fix https://github.com/kokkos/kokkos-kernels/issues/1133

ndellingwood committed:
Merge pull request #1141 from lucbv/fix_deprecation_warning

Deprecation warning: SpaceAccessibility move out of impl, see #1140

e10harvey committed:
Merge pull request #1139 from e10harvey/issue933

perf_test/blas/blas3: Update SHAs for benchmarking

masterleinad committed:
Use KOKKOS_IMPL_DO_NOT_USE_PRINTF instead of printf in kernels

uhetmaniuk committed:
Update KokkosSparse_BsrMatrix.hpp

uhetmaniuk committed:
Update KokkosSparse_BlockCrsMatrix.hpp

uhetmaniuk committed:
#1146 - Fix formatting.

kliegeois committed:
Add batched GMRES

cwpearson committed:
remove initial BlockCrsMatrix implementation

cwpearson committed:
Fixed signedness in predicates, avoid instantiation for unsupported
complex types.

ndellingwood committed:
Temporarily disable perf test KokkosSparse_sptrsv_superlu w/ Trilinos

Disabled to bypass -Werror triggered by extern C linkage within some
superlu installs

lucbv committed:
Merge pull request #1131 from lucbv/gemm_on_streams

Stream interface: adding stream support in GEMV and GEMM

lucbv committed:
Merge pull request #1147 from uhetmaniuk/1146-fix-values-size

Fix values dimension for block sparse matrices

lucbv committed:
SpMV: mismatch in #ifdef check and kernel specialization

We are checking for KOKKOSKERNELS_INST_LAYOUTLEFT to generate
the LayoutRight specialization of the algorithm. This does not
make sense, fixing the issue by modifying the #ifdef close.
This should fix issue #1148.

lucbv committed:
rocm TPLs: adding some logic to enable rocBLAS and rocSPARSE

Adding some infrastructure in cmake and tests to enable
rocBLAS and rocSPARSE and check that is compiles correctly
and that simple calls to get library versions are working.

ndellingwood committed:
Merge pull request #1150 from ndellingwood/trilinos-disable-sluperftest

Temporarily disable perf test KokkosSparse_sptrsv_superlu w/ Trilinos

lucbv committed:
TPL: adding utilities and tests for rocSPARSE

Adding similar utitlies and tests for rocSPARSE
as the ones previously added for rocBLAS.

ndellingwood committed:
develop: update version to 3.5.99 and CHANGELOG

lucbv committed:
TPL: applying clang-format to .hpp and .cpp files

masterleinad committed:
Merge branch 'develop' into use_kokkos_impl_do_not_use_printf

lucbv committed:
TPL: add a clang-format guard to avoid issues with CMake variables

lucbv committed:
rocm TPLs: changing the CMake logic a bit to have support OFF by default

Since ROCm does not ship with these libraries by default, they should
be turned OFF by default and only turned ON if the user has requested
them explicitely.

ndellingwood committed:
Add missing $ to KOKKOS_HAS_TRILINOS in sparse_sptrsv_superlu check

Fix for PR #1150

e10harvey committed:
Merge pull request #1158 from ndellingwood/update-dev-version

develop: update version to 3.5.99 and CHANGELOG

ndellingwood committed:
Merge branch 'develop' into fix-1150

ndellingwood committed:
Merge pull request #1160 from ndellingwood/fix-1150

Add missing $ to KOKKOS_HAS_TRILINOS in sparse_sptrsv_superlu check

masterleinad committed:
Disable some more tests for SYCL and OpenMPTarget

lucbv committed:
TPL rocm: adding more options to pass TPL locations

Adding KokkosKernels_ROCBLAS_ROOT and KokkosKernels_ROCSPARSE_ROOT
to the FindTPL modules as they are defined automatically by our
TPL system and should be available to users.

e10harvey committed:
Merge pull request #1145 from masterleinad/use_kokkos_impl_do_not_use_printf

Use KOKKOS_IMPL_DO_NOT_USE_PRINTF instead of printf in kernels

lucbv committed:
Merge pull request #1153 from lucbv/TPL_rocblas_rocsparse

Tpl rocblas and rocsparse

lucbv committed:
Merge pull request #1151 from lucbv/spmv_fix_layout

SpMV: mismatch in #ifdef check and kernel specialization

brian-kelley committed:
Small fixes to spgemm, and plug gaps in testing (#1159)

* Small fixes to spgemm, and plug gaps in testing

- Simplified interface: don't destroy spgemm subhandle after numeric.
Want to be able to reuse, and user is responsible for creating the
handle anyway.
- Don't default to SPGEMM_CUSPARSE unless cusparse TPL is actually
enabled.
- Don't divide by 0 when A has 0 rows.
- In testing, test the case where A has 0 rows.
- In testing, actually test default algo SPGEMM_KK, since this takes a
slightly different code path than SPGEMM_KK_MEMORY.

* In spgemm test, rename dimensions to m/k/n

cwpearson committed:
remove public: in struct

cwpearson committed:
remove commented-out debug messages

cwpearson committed:
add kk_is_gpu_mem_space()

cwpearson committed:
don't refer to identifiers that are only defined when KOKKOS_ENABLE_CUDA

cwpearson committed:
only do local typedefs when used

lucbv committed:
Fixing warning for auto-tester

removing an unused typedef that bothers the auto-tester.

lucbv committed:
Removing one more typedef

Did not realize that the previous typedef would also need to go.

kliegeois committed:
Update of the GMRES PR

lucbv committed:
using namespace should be scoped to prevent name clashes, see issue #1170

Introducing a whole namespace in a public is dangerous due to the
potential for name clashes with other libraries.
Removing such usage from trmm, trsm and trtri resolves build issues
observed by users in Trilinos.

lucbv committed:
Merge pull request #1171 from lucbv/Fix_using_namespace

using namespace should be scoped to prevent name clashes, see issue #1170

lucbv committed:
Merge pull request #1090 from cwpearson/feature/tensor-core-spmv

Experimental tensor-core SpMV for BsrMatrix

lucbv committed:
SpMV: fix catch all case to avoid compiler warnings

As usually removing names of parameters from function
signature.

lucbv committed:
SpMV: applying clang-format

ndellingwood committed:
KokkosSparse_spmv_bsrmatrix_impl: remove unused typedef

Removed unused Ordinal typdef to prevent -Werror compile errors with
clang compilers

lucbv committed:
Merge pull request #1179 from lucbv/spmv_clean_ups

SpMV: fix catch all case to avoid compiler warnings

lucbv committed:
Merge pull request #1180 from ndellingwood/remove-unused-typedef

KokkosSparse_spmv_bsrmatrix_impl: remove unused typedef

lucbv committed:
Merge pull request #1162 from masterleinad/disable_tests_sycl_ompt

Use KOKKOS_IMPL_DO_NOT_USE_PRINTF instead of printf in kernels

lucbv committed:
Fixing a couple of small issue with tensor core spmv

The nightly tests are still reporting some issues.
This PR should take care of some of them!

brian-kelley committed:
Fix #threads setting in pcg for OpenMP

./sparse_pcg --openmp N ... should set the number of threads to N,
but it was setting this to 0 (causing a crash in parallel_scan much
later).

ndellingwood committed:
cm_test_all_sandia: add clang/10.0.1 option for blake

lucbv committed:
Merge pull request #1182 from brian-kelley/FixGSOpenMP

Fix #threads setting in pcg for OpenMP

masterleinad committed:
Add nightly CI check for SYCL

masterleinad committed:
Disable HIP for now

masterleinad committed:
Restrict to CPU

masterleinad committed:
Add Build & Run stage

masterleinad committed:
Run on GPU

brian-kelley committed:
Check that ordinal types are signed

This is required by several algorithms and is true of the possible ETI
types. Having an unsigned ordinal gives warnings now. It might take some
work to make it fully support unsigned.

masterleinad committed:
Fix assert in src/sparse/KokkosSparse_BsrMatrix.hpp

lucbv committed:
SpMV: fixing issue with variable usage

was using c instead of pattern...
build on Weaver shows this should now work fine.

masterleinad committed:
Fix tan

masterleinad committed:
Rather return directly

lucbv committed:
Merge pull request #1185 from lucbv/fix_nightly_builds

Fixing a couple of small issue with tensor core spmv

lucbv committed:
Merge pull request #1189 from masterleinad/fix_assert

Fix assert in src/sparse/KokkosSparse_BsrMatrix.hpp

lucbv committed:
Merge pull request #1188 from brian-kelley/CheckOrdinalSigned

Check that ordinal types are signed

masterleinad committed:
More changes to Kokkos_ArithTraits.hpp

lucbv committed:
Merge pull request #1183 from ndellingwood/update-testall-blake

cm_test_all_sandia: add clang/10.0.1 option for blake

masterleinad committed:
Disable gemv_double test

masterleinad committed:
Disable more unit tests

masterleinad committed:
Add HIP build again

masterleinad committed:
Try disabling the ArithTraits tests

masterleinad committed:
Timeout after 3 hours

masterleinad committed:
Update the HIP Dockerfile to be similar to the Kokkos one

e10harvey committed:
.github/workflows:

  - Add EOL control character to regex.

masterleinad committed:
Don't run unit tests for SYCL yet

lucbv committed:
Merge pull request #1190 from masterleinad/nightly-sycl

Add nightly CI check for SYCL

e10harvey committed:
example/gmres: Use fabs

  - Fixes #1172

sparse: Fix unused vars in bsrmatrix impl

e10harvey committed:
scripts/cm_test_all_sandia: add clang/10.0.1 spot check

Commit:
Copy SpGEMM files for reference diff

e10harvey committed:
Merge pull request #1198 from e10harvey/issue956

scripts/cm_test_all_sandia: add clang/10.0.1 spot check

Commit:
ETI specializations for block SpGEMM

brian-kelley committed:
Add rocBLAS GEMV wrapper

Fix GEMV for complex/LayoutRight: this isn't supported by cuBLAS or
rocBLAS, so call the fallback option instead of just throwing.

Add HIP support to GEMV perftest

brian-kelley committed:
GEMV perf test: fix useSerial condition

Run with serial if no parallel backend is selected. Also fix layout condition for Threads.

brian-kelley committed:
GEMV mode C, LayoutRight: dont call host blas

brian-kelley committed:
Allow choosing coloring algorithm in multicolor GS (#1199)

* Allow choosing coloring algorithm in multicolor GS

(both point and cluster)

* Diag-dominant matrix: alway make room for diagonal

Commit:
Unit test for block SpGEMM

Commit:
Implementation of default block SpGEMM algorithm

Commit:
Implementation of serial/debug block SpGEMM

e10harvey committed:
examples/gmres: use namespace std

lucbv committed:
Merge pull request #1201 from brian-kelley/RocBlasGemv

Add rocBLAS GEMV wrapper

e10harvey committed:
examples/gmres: use namespace Issue1172WorkAround

ndellingwood committed:
Merge pull request #1197 from e10harvey/issue1172

example/gmres: Use fabs

e10harvey committed:
example/gmres: Specify half_t namespace

e10harvey committed:
example/gmres: Use complex double

ndellingwood committed:
Merge pull request #1208 from e10harvey/issue1157-2

example/gmres: Specify half_t namespace

lucbv committed:
Merge pull request #1155 from kliegeois/GMRES

Add Batched CG and Batched GMRES

mzuzek commented:
Closing this internal version - please comment on regular PR https://github.com/kokkos/kokkos-kernels/pull/1099
